aehmc/termination.py

```diff
@@ -0,0 +1,128 @@
+from typing import Callable, Tuple
+
+import aesara
+import aesara.tensor as aet
+import numpy as np
+from aesara.ifelse import ifelse
+from aesara.scan.utils import until
+from aesara.tensor.var import TensorVariable
+
+
+def iterative_uturn(is_turning_fn: Callable):
+    """U-Turn termination criterion to check reversiblity while expanding
+    the trajectory in a multiplicative way.
+
+    The code follows the implementation in Numpyro [0]_, which is equivalent to
+    that in TFP [1]_.
+
+    References
+    ----------
+    .. [0]: Phan, Du, Neeraj Pradhan, and Martin Jankowiak. "Composable effects
+            for flexible and accelerated probabilistic programming in NumPyro." arXiv
+            preprint arXiv:1912.11554 (2019).
+    .. [1]: Lao, Junpeng, et al. "tfp. mcmc: Modern markov chain monte carlo
+            tools built for modern hardware." arXiv preprint arXiv:2002.01184 (2020).
+
+    """
+
+    def new_state(
+        position: TensorVariable, max_num_doublings: int
+    ) -> Tuple[TensorVariable, TensorVariable, int, int]:
+        num_dims = position.ndim
+        return (
+            aet.zeros((max_num_doublings, num_dims)),
+            aet.zeros((max_num_doublings, num_dims)),
+            0,
+            0,
+        )
+
+    def update(
+        state: Tuple, momentum_sum: TensorVariable, momentum: TensorVariable, step: int
+    ):
+        momentum_ckpt, momentum_sum_ckpt, *_ = state
+        idx_min, idx_max = _find_storage_indices(step)
+        momentum_ckpt = aet.where(
+            aet.eq(step % 2, 0),
+            aet.set_subtensor(momentum_ckpt[idx_max], momentum),
+            momentum_ckpt,
+        )
```

Comment:
@brandonwillard could you help me understand the following error?

<details><summary>Error message</summary>
<p>

```
E           ValueError: Dimension 0 in Rebroadcast's input was supposed to be 1 (got 3 instead)
E           Apply node that caused the error: Rebroadcast{1}(TensorConstant{(3,) of 1.0})
E           Toposort index: 4
E           Inputs types: [TensorType(float64, vector)]
E           Inputs shapes: [(3,)]
E           Inputs strides: [(8,)]
E           Inputs values: [array([1., 1., 1.])]
E           Outputs clients: [[IncSubtensor{InplaceSet;int32}(Alloc.0, Rebroadcast{1}.0, ScalarFromTensor.0)]]
E           
E           Backtrace when the node is created (use Aesara flag traceback__limit=N to make it longer):
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
E               self.ihook.pytest_pyfunc_call(pyfuncitem=self)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
E               return self._hookexec(self, self.get_hookimpls(), kwargs)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
E               return self._inner_hookexec(hook, methods, kwargs)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
E               self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
E               res = hook_impl.function(*args)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
E               result = testfunction(**testargs)
E             File "/home/remi/projects/ampersand/aehmc/tests/test_termination.py", line 52, in test_termination_update
E               updated = update(termination_state, momentum_sum, momentum, step)
E             File "/home/remi/projects/ampersand/aehmc/aehmc/termination.py", line 46, in update
E               aet.set_subtensor(momentum_ckpt[idx_max], momentum),
```

</p>
</details>

Reply:
I don't know what's generating that `Rebroadcast`, so I would try to determine that first.  The `set_subtensor` (and/or the `*Subtensor*` implied by `momentum_ckpt[idx_max]`) is apparently involved, so try printing that sub-graph with `aesara.dprint`.

Otherwise, it looks like the shapes/dimension of the inputs are incompatible with the graph, because `Rebroadcast{1}` is essentially a check that makes sure the first dimension is broadcastable (i.e. its size is equal to 1), but it has been given an array with a dimension that isn't `1` (i.e. `np.array([1., 1., 1.])`).